### ISSUE TITLE
If ProxySelector getDefault() throws AccessControlException, we should default to NullProxySelector

### ISSUE THREAD

**Issue by theonecf** [square/okhttp#5074]
Hi there, I am trying to use okhttp but run into security permissions. It failed at ProxySelector getDefault with SecurityManager checkPermission on "getProxySelector". 

For any case if getDefault throws exception, I think we should fall into NullProxySelector.

 ` this.proxySelector = ProxySelector.getDefault();`
  ` if (this.proxySelector == null) {`
 `               this.proxySelector = new NullProxySelector();`
   `         }`

**Comment by swankjesse** [comment]
Are you really sure you need a SecurityManager?

**Comment by theonecf** [comment]
> Are you really sure you need a SecurityManager?

Yes, even we can turn off security manager in DEV but in PROD we need it.

**Comment by swankjesse** [comment]
Tell me more. If your production environment relies on that security manager for security, I have bad news.

**Comment by theonecf** [comment]
> Tell me more. If your production environment relies on that security manager for security, I have bad news.

We have a java jar deployed in Apigee Edge(Cloud based). Apigee itself has some java security permissions setup. https://docs.apigee.com/api-platform/reference/java-permission-reference Unfortunately, we cannot override any security policy there. 

I think we at least need a way to construct OkHttpClient.java with no proxy in okhttp lib.

**Comment by swankjesse** [comment]
Gotcha. Hopefully they aren't using that for real security, as it doesn't offer any.

Does providing a ProxySelector explicitly work around the problem?

**Comment by theonecf** [comment]
> Gotcha. Hopefully they aren't using that for real security, as it doesn't offer any.
> 
> Does providing a ProxySelector explicitly work around the problem?
Hmm, not quite sure. We don't have control over Apigee deployment JVM .

**Comment by theonecf** [comment]
> Gotcha. Hopefully they aren't using that for real security, as it doesn't offer any.
> 
> Does providing a ProxySelector explicitly work around the problem?

**Comment by swankjesse** [comment]
You should be able to configure a ProxySelector manually in your OkHttpClient.Builder().

**Comment by mkw** [comment]
Unfortunately, we are using OkHttp indirectly via another library that we do not control, so we cannot modify the construction of the OkHttp client. Given how widely OkHttp is wrapped by other libraries, I would think that this is a pretty common use case.

When the proxy selected permission is denied by a security manager, it is likely that the only reasonable option is a null proxy selector.  I would argue that a warning in the logs is more appropriate than failure.

**Comment by yschimke** [comment]
It's the public API, so maybe worth supporting?  Jesse, would you take a PR?

https://docs.oracle.com/javase/7/docs/api/java/net/ProxySelector.html#getDefault()

Throws:
    SecurityException - If a security manager has been installed and it denies NetPermission("getProxySelector")

**Comment by swankjesse** [comment]
That whole situation sounds just painful. We can probably alleviate some of it with a catch clause. But you should avoid the harmful SecurityManager and you should get your libraries to let you configure the OkHttpClient that they use.

**Comment by theonecf** [comment]
> That whole situation sounds just painful. We can probably alleviate some of it with a catch clause. But you should avoid the harmful SecurityManager and you should get your libraries to let you configure the OkHttpClient that they use.

Unfortunately, we do not control the provided JVM and cannot avoid the SecurityManager. Also we use OkHttpClient indirectly from another library as mentioned above.

**Comment by yschimke** [comment]
https://github.com/square/okhttp/pull/5332
